experimental/sql/src/projection_optimizations.py

```diff
@@ -0,0 +1,191 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from xdsl.ir import Operation, MLContext, Region, Block
+from typing import List, Type, Optional
+from xdsl.dialects.builtin import ArrayAttr, StringAttr, ModuleOp, IntegerAttr
+
+from xdsl.pattern_rewriter import (RewritePattern, GreedyRewritePatternApplier,
+                                   PatternRewriteWalker, PatternRewriter,
+                                   op_type_rewrite_pattern)
+
+import dialects.rel_alg as RelAlg
+
+# This file contains several rewrites on the rel_alg level, all of which concern
+# projections. Currently, there are three rewriters. The first one introduces a
+# projection before any operator, the second one fuses subsequent projections,
+# and the third one removes projetions that are just identity mappings.
+
+
+@dataclass
+class ProjectionOptimizer(RewritePattern):
+
+  def flatten(self, l: list[list[str]]) -> list[str]:
+    return [elem for sublist in l for elem in sublist]
+
+  def find_cols_in_expr(self, op: RelAlg.Expression) -> list[str]:
+    """
+    Returns the list of cols that are used in this expression or its children.
+    """
+    if isinstance(op, RelAlg.Column):
+      return [op.col_name.data]
+    else:
+      return self.flatten([self.find_cols_in_expr(r.op) for r in op.regions])
+
+  def find_cols_upstream(self, op: RelAlg.Operator) -> list[str]:
+    """
+    Returns the list of cols that are used in this operation or its parents. If
+    the list contains None, a ModuleOp (so the output) was hit, showing that all
+    columns need to be kept.
+    """
+    if isinstance(op, ModuleOp):
+      return [None]
+    if isinstance(op, RelAlg.Select):
+      return self.flatten([
+          self.find_cols_in_expr(o) for o in op.predicates.ops
+      ]) + self.find_cols_upstream(op.parent_op())
+    if isinstance(op, RelAlg.Project):
+      return self.flatten(
+          [self.find_cols_in_expr(o) for o in op.projections.ops])
+    if isinstance(op, RelAlg.Aggregate):
+      return [s.data for s in op.col_names.data]
+
+  def find_schema(self, op: RelAlg.Operator) -> list[str]:
```

Comment:
I am wondering whether this isn't so fundamental that it could be a property/member function of `class Operator`.